tests/Giraffe.Tests/HttpHandlerTests.fs

```diff
@@ -623,48 +609,51 @@ let ``GET "/api/users" returns "users"`` () =
     ctx.Response.Body <- new MemoryStream()
     let expected = "users"
 
-    let result =
+    task {
+        let! result = app next ctx
 
-        app next ctx
-        |> Async.RunSynchronously
-
-    match result with
-    | None -> assertFailf "Result was expected to be %s" expected
-    | Some ctx ->
-        let body = getBody ctx
-        Assert.Equal(expected, body)
+        match result with
+        | None -> assertFailf "Result was expected to be %s" expected
+        | Some ctx ->
+            let body = getBody ctx
+            Assert.Equal(expected, body)
+    }
 
 [<Fact>]
 let ``GET "/api/test" returns "test"`` () =
-    let ctx = Substitute.For<HttpContext>()
-    let app =
-        GET >=> choose [
-            route "/"    >=> text "Hello World"
-            route "/foo" >=> text "bar"
-            subRoute "/api" (
-                choose [
-                    route ""       >=> text "api root"
-                    route "/admin" >=> text "admin"
-                    route "/users" >=> text "users" ] )
-            route "/api/test" >=> text "test"
-            setStatusCode 404 >=> text "Not found" ]
-
-    ctx.Items.Returns (new Dictionary<obj,obj>() :> IDictionary<obj,obj>) |> ignore
-    ctx.Request.Method.ReturnsForAnyArgs "GET" |> ignore
-    ctx.Request.Path.ReturnsForAnyArgs (PathString("/api/test")) |> ignore
-    ctx.Response.Body <- new MemoryStream()
-    let expected = "test"
-
-    let result =
 
-        app next ctx
-        |> Async.RunSynchronously
+    task {
+        let ctx = Substitute.For<HttpContext>()
 
-    match result with
-    | None -> assertFailf "Result was expected to be %s" expected
-    | Some ctx ->
-        let body = getBody ctx
-        Assert.Equal(expected, body)
+        let app =
+            GET >=> choose [
+                route "/"    >=> text "Hello World"
+                //route "/foo" >=> text "bar"
+                subRoute "/api" (
+                    choose [
+                        route ""       >=> text "api root"
+                        route "/admin" >=> text "admin"
+                        route "/users" >=> text "users" ] )
+                route "/foo" >=> text "bar"
+                route "/api/test" >=> text "test" //repositioned
```

Comment:
Sorry, should have been more clear, the try/finally works a little differently in TaskBuilder.fs then did before (i didn't change it) and while I was going to fix up & correct, I didn't see why we were using try/finally to "Hack" the functionality (so that we could basically run a finalizer when task dropped out of scope etc), and with there being no need to use try/finally, I fixed it to use proper logic flow, stripping back stem on failed subroute match.

TaskBuilder.fs runs finally after there is a full `Return` value, not a bound task, not too difficult to tweak to how async does it.

I will fix try/finally in taskbuilder as I re-write/rebuild it but I really not sure why subroute was built the way it was using try/finally, that construct is for code with likely exceptions that need resources cleaned up (disposables etc), bizarre how it is being used in `handlerWithRoutedPath`.

If replicating prior try/finally function flow is a priority, I can tweak before we release?

NB: I think we can take it that basically, all Tasks are Hot Tasks as the framework is driving/starting them, and nearly all Task Async method calls are hot too ... when using async, it was all Hot Async as we were scheduling async based on the hot tasks (double scheduling), nothing was really being created in a cold state to be run later.